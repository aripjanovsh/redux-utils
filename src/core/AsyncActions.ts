import { Action, AnyAction } from "redux";

export interface PerformAction<M = any> extends Action {
  readonly meta: M;
}

export interface FulfillAction<P, M = any> extends PerformAction<M> {
  readonly payload: P;
}

export interface RejectAction<M = any> extends PerformAction<M> {
  readonly error: true;
  readonly payload: Error;
}

export function isPerformAction(action: AnyAction): action is PerformAction {
  return Boolean(
    action &&
      action.meta !== undefined &&
      action.error === undefined &&
      action.payload === undefined,
  );
}

export function createPerformAction<M>(
  type: string,
  meta: M,
): PerformAction<M> {
  return { meta, type };
}

export function isFulfillAction(
  action: AnyAction,
): action is FulfillAction<any> {
  return Boolean(
    action &&
      action.meta !== undefined &&
      action.error === undefined &&
      action.payload !== undefined,
  );
}

export function createFulfillAction<P, M = any>(
  type: string,
  payload: P,
  meta: M,
): FulfillAction<P, M> {
  return { meta, type, payload };
}

export function isRejectAction(action: AnyAction): action is RejectAction {
  return Boolean(
    action &&
      action.error === true &&
      action.meta !== undefined &&
      action.payload !== undefined,
  );
}

export function createRejectAction<M = any>(
  type: string,
  error: Error,
  meta: M,
): RejectAction<M> {
  return { meta, type, error: true, payload: error };
}
