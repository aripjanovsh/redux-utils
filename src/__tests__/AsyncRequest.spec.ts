import { Action } from "redux";

import { FulfillAction, PerformAction, RejectAction } from "../AsyncActions";
import {
  createAsyncRequestReducer,
  getInitialAsyncRequest,
} from "../AsyncRequest";

describe("AsyncRequest", () => {
  describe("getInitialAsyncRequest", () => {
    test("initial value", () => {
      expect(getInitialAsyncRequest()).toMatchSnapshot();
    });
  });

  describe("createAsyncRequestReducer", () => {
    const perform = "perform";
    const fulfill = "fulfill";
    const reject = "reject";
    const reset = "reset";

    test("creation", () => {
      const reducer = createAsyncRequestReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      expect(typeof reducer).toBe("function");
    });

    test("perform action", () => {
      const reducer = createAsyncRequestReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("Async Error");
      const action: PerformAction = { meta: {}, type: perform };

      expect(
        reducer(
          {
            submitted: false,
            submitting: false,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        submitted: true,
        submitting: true,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            response: 10,
            submitted: true,
            submitting: false,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 10,
        submitted: true,
        submitting: true,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            submitted: true,
            submitting: false,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        error,
        submitted: true,
        submitting: true,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            response: 10,
            submitted: true,
            submitting: false,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        error,
        response: 10,
        submitted: true,
        submitting: true,
        submitFailed: false,
      });
    });

    test("fulfill action", () => {
      const reducer = createAsyncRequestReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("Async Error");
      const action: FulfillAction<number> = {
        meta: {},
        payload: 30,
        type: fulfill,
      };

      expect(
        reducer(
          {
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 30,
        submitted: true,
        submitting: false,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            response: 10,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 30,
        submitted: true,
        submitting: false,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            submitted: true,
            submitting: false,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 30,
        submitted: true,
        submitting: false,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            response: 10,
            submitted: true,
            submitting: false,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 30,
        submitted: true,
        submitting: false,
        submitFailed: false,
      });
    });

    test("reject action", () => {
      const reducer = createAsyncRequestReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const stateError = new Error("AsyncError: State.");
      const actionError = new Error("AsyncError: Action.");
      const action: RejectAction = {
        meta: {},
        error: true,
        type: reject,
        payload: actionError,
      };

      expect(
        reducer(
          {
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        error: actionError,
        submitted: true,
        submitting: false,
        submitFailed: true,
      });

      expect(
        reducer(
          {
            response: 10,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 10,
        error: actionError,
        submitted: true,
        submitting: false,
        submitFailed: true,
      });

      expect(
        reducer(
          {
            error: stateError,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        error: actionError,
        submitted: true,
        submitting: false,
        submitFailed: true,
      });

      expect(
        reducer(
          {
            response: 10,
            error: stateError,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 10,
        error: actionError,
        submitted: true,
        submitting: false,
        submitFailed: true,
      });
    });

    test("reset action", () => {
      const reducer = createAsyncRequestReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("AsyncError: State.");
      const action: Action = { type: reset };

      expect(
        reducer(
          {
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        submitted: false,
        submitting: false,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            response: 10,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        submitted: false,
        submitting: false,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        submitted: false,
        submitting: false,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            response: 10,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        submitted: false,
        submitting: false,
        submitFailed: false,
      });
    });

    test("unknown action", () => {
      const reducer = createAsyncRequestReducer<number>(
        perform,
        fulfill,
        reject,
        reset,
      );

      const error = new Error("AsyncError: State.");
      const action: Action = { type: "FOO" };

      expect(
        reducer(
          {
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        submitted: true,
        submitting: true,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            response: 10,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        response: 10,
        submitted: true,
        submitting: true,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        error,
        submitted: true,
        submitting: true,
        submitFailed: false,
      });

      expect(
        reducer(
          {
            error,
            response: 10,
            submitted: true,
            submitting: true,
            submitFailed: false,
          },
          action,
        ),
      ).toEqual({
        error,
        response: 10,
        submitted: true,
        submitting: true,
        submitFailed: false,
      });
    });
  });
});
