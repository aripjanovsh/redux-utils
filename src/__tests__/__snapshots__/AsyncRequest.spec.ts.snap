// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncRequest getInitialAsyncRequest initial value 1`] = `
Object {
  "requestFailed": false,
  "requested": false,
  "requesting": false,
}
`;
